### PR TITLE
[Snappi] Added multi dut PFC lossy test case

### PR DESCRIPTION
**Pull request by vikumarks**
### Description of PR
- Testing PFC lossy priorities behavior when sending external pfc pause frames in modular chassis which has multi-asic 

Summary:
Fixes # (issue)

### Type of change
- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ x] Test case(new/improvement)

### Back port request
- [ ] 201911
- [ ] 202012
- [ ] 202205

#### What is the motivation for this PR?
To test PFC lossy behavior in multi-dut/multi-asic scenario. 
#### How did you do it?
The existing RDMA cases for single dut has been modified to accommodate T2 topology, where separate fixtures has been added that takes into account the chassis type and asic choice as user input, so that the script supports both T1 and T2 topology
#### How did you verify/test it?
Tested against voq chassis with SONiC installed in it 
#### Any platform specific information?
Tested against J2C+ 
#### Supported testbed topology if it's a new test case?
T2, T1 & T0 
### Documentation
https://github.com/sonic-net/sonic-mgmt/blob/master/docs/testplan/pfc/PFC_PAUSE_LOSSY_README.md

### CODE REVIEW

**Comment by kamalsahu0001** [comment]
Testplan for PFC Lossy testcase is same for both singledut and multidut
https://github.com/sonic-net/sonic-mgmt/blob/master/docs/testplan/pfc/PFC_PAUSE_LOSSY_README.md

**Comment by developfast** [comment]
@vikumarks We've recently implemented updates to the Snappi tests, relocating all test cases to the following location: https://github.com/sonic-net/sonic-mgmt/tree/master/tests/snappi_tests. Kindly make the necessary adjustments to reflect this change.

Furthermore, I kindly request you take a look at the pull request at https://github.com/sonic-net/sonic-mgmt/pull/8660. Please ensure that the PFC tests for multi DUT environments align harmoniously with the existing PFC tests that have been refactored.

**Comment by developfast** [comment]
Please add the appropriate new parameters added to this PR to `snappi_test_params`.

**Review by developfast — CHANGES_REQUESTED** [review]
@vikumarks Can you address the pre-commit failures? One more thing, for these multidut cases, can you move them to a multidut folder inside the snappi_tests directory? For instance, for pfc multidut cases, can you put it in a folder like `tests/snappi_tests/multidut/pfc/...`? This is to make organisation easier when we are running these tests on our nightly testbeds.

**Comment by developfast** [comment]
Can you list all dependent PRs i.e. PR numbers with a # for this PR to be merged? This will make tracking easier

**Comment by vikumarks** [comment]
> Please add the appropriate new parameters added to this PR to `snappi_test_params`.
@developfast , Params are part of PR #8186

**Comment by vikumarks** [comment]
> @vikumarks Can you address the pre-commit failures? One more thing, for these multidut cases, can you move them to a multidut folder inside the snappi_tests directory? For instance, for pfc multidut cases, can you put it in a folder like `tests/snappi_tests/multidut/pfc/...`? This is to make organisation easier when we are running these tests on our nightly testbeds.

@developfast, moved the cases to multidut folder

**Review by vmittal-msft — APPROVED** [review]
Looks ok to me. Can you please add more comments to make code more readable? Also, please share test results and SW version details.

@baiwei0427 @developfast please see if this looks ok.

File: tests/snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+        line_card_choice: Line card choice to be mentioned in the variable.py file
+        linecard_configuration_set : Line card classification, (min 1 or max 2  hostnames and asics to be given)
+
+    Returns:
+        N/A
+    """
+    if line_card_choice not in linecard_configuration_set.keys():
+        assert False, "Invalid line_card_choice value passed in parameter"
+
+    if (len(linecard_configuration_set[line_card_choice]['hostname']) == 2):
```
Comment (wenyiz2021) [inline tests/snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py:50]: should this be >= 2?

File: tests/snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
```
Comment (wenyiz2021) [inline tests/snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py:127]: same here?

File: tests/snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
```
Comment (wenyiz2021) [inline tests/snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py:208]: same here

File: tests/snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+        linecard_configuration_set : Line card classification, (min 1 or max 2  hostnames and asics to be given)
+
+    Returns:
+        N/A
+    """
+
+    if line_card_choice not in linecard_configuration_set.keys():
+        assert False, "Invalid line_card_choice value passed in parameter"
+
+    if (len(linecard_configuration_set[line_card_choice]['hostname']) == 2):
```
Comment (wenyiz2021) [inline tests/snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py:299]: same here

**Review by wenyiz2021 — CHANGES_REQUESTED** [review]
please see my comments

**Review by developfast — CHANGES_REQUESTED** [review]

File: tests/snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py (hunk 1 above)
Comment (vikumarks) [reply tests/snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py]: Fixed

File: tests/snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py (hunk 2 above)
Comment (vikumarks) [reply tests/snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py]: Fixed

File: tests/snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py (hunk 3 above)
Comment (vikumarks) [reply tests/snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py]: Fixed

File: tests/snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py (hunk 4 above)
Comment (vikumarks) [reply tests/snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py]: Fixed

File: tests/snappi_tests/pfc/test_multidut_pfc_pause_lossy_with_snappi.py [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+from tests.snappi_tests.variables import config_set, line_card_choice
+from files.multidut_helper import run_pfc_test
+from tests.common.reboot import reboot
+from tests.common.utilities import wait_until
+import logging
+import random
+from tests.common.snappi_tests.snappi_test_params import SnappiTestParams
+logger = logging.getLogger(__name__)
+
+pytestmark = [pytest.mark.topology('snappi')]
```
[the remark this answers is not in the document]
Comment (vikumarks) [reply tests/snappi_tests/pfc/test_multidut_pfc_pause_lossy_with_snappi.py]: Fixed

File: tests/snappi_tests/pfc/test_multidut_pfc_pause_lossy_with_snappi.py [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+@pytest.mark.parametrize('line_card_choice', [line_card_choice])
+@pytest.mark.parametrize('linecard_configuration_set', [config_set])
+def test_pfc_pause_single_lossy_prio(snappi_api,                # noqa: F811
+                                     conn_graph_facts,          # noqa: F811
+                                     fanout_graph_facts,        # noqa: F811
+                                     duthosts,
+                                     line_card_choice,
+                                     linecard_configuration_set,
+                                     get_multidut_snappi_ports  # noqa: F811
+                                     ):
```
[the remark this answers is not in the document]
Comment (vikumarks) [reply tests/snappi_tests/pfc/test_multidut_pfc_pause_lossy_with_snappi.py]: Fixed

**Comment by vmittal-msft** [comment]
@vikumarks please take care of comments. @baiwei0427  @developfast  is this ok to merge ?

**Review by developfast — CHANGES_REQUESTED** [review]
All plain `assert` cases need to be changed to either `pytest_assert` or `pytest_require`

File: tests/snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+        N/A
+    """
+    if line_card_choice not in linecard_configuration_set.keys():
+        assert False, "Invalid line_card_choice value passed in parameter"
+
+    if (len(linecard_configuration_set[line_card_choice]['hostname']) >= 2):
+        dut_list = random.sample(duthosts, 2)
+        duthost1, duthost2 = dut_list
+    elif (len(linecard_configuration_set[line_card_choice]['hostname']) == 1):
+        dut_list = [dut for dut in duthosts if linecard_configuration_set[line_card_choice]['hostname'] == [dut.hostname]]  # noqa: E501
```
[the remark this answers is not in the document]
Comment (vikumarks) [reply tests/snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py]: Fixed

File: tests/snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py [hunk 8]
```diff
@@ -0,0 +1,349 @@
...
+        duthosts (pytest fixture): list of DUTs
+        prio_dscp_map (pytest fixture): priority vs. DSCP map (key = priority).
+        line_card_choice: Line card choice to be mentioned in the variable.py file
+        linecard_configuration_set : Line card classification, (min 1 or max 2  hostnames and asics to be given)
+
+    Returns:
+        N/A
+    """
+    if line_card_choice not in linecard_configuration_set.keys():
+        assert False, "Invalid line_card_choice value passed in parameter"
```
[the remark this answers is not in the document]
Comment (vikumarks) [reply tests/snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py]: Fixed

File: tests/snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py [hunk 9]
```diff
@@ -0,0 +1,349 @@
...
+
+    if (len(linecard_configuration_set[line_card_choice]['hostname']) >= 2):
+        dut_list = random.sample(duthosts, 2)
+        duthost1, duthost2 = dut_list
+    elif (len(linecard_configuration_set[line_card_choice]['hostname']) == 1):
+        dut_list = [dut for dut in duthosts if linecard_configuration_set[line_card_choice]['hostname'] == [dut.hostname]]  # noqa: E501
+        duthost1, duthost2 = dut_list[0], dut_list[0]
+    else:
+        assert False, "Hostname can't be an empty list"
```
[the remark this answers is not in the document]
Comment (vikumarks) [reply tests/snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py]: Fixed

**Comment by vmittal-msft** [comment]
@developfast please approve if ok to merge.

**Review by developfast — APPROVED** [review]
lgtm

**Comment by developfast** [comment]
@yxieca can you pls help merge?

**Comment by yxieca** [comment]
@vikumarks can you fill the PR template "how did you do it" section with a brief description of the fix?

**Comment by vikumarks** [comment]
@yxieca Filled the missing "how did you do it" section

**Comment by yxieca** [comment]
@wenyiz2021 please assess if the issue you raised had been addressed?

**Comment by vikumarks** [comment]
> @yxieca Filled the missing "how did you do it" section

PR Template has been completed as requested , Please merge.

**Review by judyjoseph — APPROVED** [review]

**Comment by vikumarks** [comment]
@yxieca can you please merge this PR

**Review by wenyiz2021 — COMMENTED** [review]
LGTM

**Pull request merged**